massa-cipher/src/decrypt.rs

```diff
@@ -2,34 +2,38 @@
 
 use aes_gcm_siv::aead::{Aead, NewAead};
 use aes_gcm_siv::{Aes256GcmSiv, Key, Nonce};
-use massa_hash::Hash;
+use pbkdf2::{
+    password_hash::{PasswordHasher, SaltString},
+    Pbkdf2,
+};
 
-use crate::constants::NONCE_SIZE;
+use crate::constants::{NONCE_SIZE, SALT_SIZE};
 use crate::error::CipherError;
 
 /// Decryption function using AES-GCM-SIV cipher.
 ///
 /// Read `lib.rs` module documentation for more information.
-pub fn decrypt(password: &str, data: &[u8]) -> Result<Vec<u8>, CipherError> {
-    let cipher = Aes256GcmSiv::new(Key::from_slice(
-        Hash::compute_from(password.as_bytes()).to_bytes(),
-    ));
-    let nonce = Nonce::from_slice(data.get(..NONCE_SIZE).ok_or_else(|| {
+pub fn decrypt(password: &str, mut data: &[u8]) -> Result<Vec<u8>, CipherError> {
+    let salt_data = data.take(..SALT_SIZE).ok_or_else(|| {
+        CipherError::DecryptionError(
+            "wallet file truncated: salt missing or incomplete".to_string(),
+        )
+    })?;
+    let salt = SaltString::new(std::str::from_utf8(salt_data)?)
+        .map_err(|e| CipherError::DecryptionError(e.to_string()))?;
+    let password_hash = Pbkdf2
+        .hash_password(password.as_bytes(), &salt)
+        .map_err(|e| CipherError::DecryptionError(e.to_string()))?
+        .hash
+        .unwrap();
+    let cipher = Aes256GcmSiv::new(Key::from_slice(password_hash.as_bytes()));
+    let nonce = Nonce::from_slice(data.take(..NONCE_SIZE).ok_or_else(|| {
```

Comment:
It doesn't take the salt again I've tested it. Also I'm not sure I understand what you mean because the slice _is_ mutable.

Reply:
haha the joys of reviewing on the phone...

But still it feels strange that the encryption function empties the data you give it and leaves it empty after it is called.
I would vouch for using get() instead of take() so that we don't need to pass mutable references

Reply:
Done